### PR TITLE
feat(cli): add support for poll watcher

### PR DESCRIPTION
**Pull request by amribm**
Closes https://github.com/vectordotdev/vector/issues/21111

### CODE REVIEW

File: src/app.rs
Comment (jorgehermo9) [inline src/app.rs:491]: you can use `if let` statements to get rid of the unwrap

https://doc.rust-lang.org/rust-by-example/flow_control/if_let.html

File: src/app.rs
Comment (jorgehermo9) [inline src/app.rs:550]: Please, take a look into my comment about this in your closed PR

https://github.com/vectordotdev/vector/pull/21258#discussion_r1756392444

File: src/config/watcher.rs
Comment (jorgehermo9) [inline src/config/watcher.rs:9]: Why this rename? I think it would be better to normalize create custom aliases as much as possible. In this case, there is no conflict with that trait name

Comment (jorgehermo9) [reply src/config/watcher.rs:9]: Oh I see you have a `Watcher` enum.. I don't know if we could find a better name for it and not use an alias.. maybe specifying the notify watcher with `notify::Watcher`?

File: src/config/watcher.rs
Comment (jorgehermo9) [inline src/config/watcher.rs:41]: you can get rid of the dynamic dispatch if you return the `Watcher` enum here.

You can implement the trait `From<RecommendedWatcher> for Watcher` and `From<PollingWatcher> Watcher` to simplify the creation of those here. You could then just call `.into`

Read this if you are not familiar with that https://doc.rust-lang.org/std/convert/trait.From.html

Comment (jorgehermo9) [reply src/config/watcher.rs:41]: You can then implement a `watch` method in `Watcher` that relies on the underlying `notify::Watcher` variants

File: src/app.rs
Comment (amribm) [reply src/app.rs:491]: okay

File: src/config/watcher.rs
Comment (amribm) [reply src/config/watcher.rs:41]: great i'll do this.

File: src/app.rs
Comment (amribm) [reply src/app.rs:550]: Can we put this as `Recommended` or any thing that comes to ur mind?

File: src/config/watcher.rs
Comment (jorgehermo9) [inline src/config/watcher.rs:141]: Should we allow the user to specify `compare_contents`? https://docs.rs/notify/latest/notify/struct.Config.html#method.with_compare_contents
maybe another argument? Im afraid we are adding too much arguments for this though

@jszwedko

File: src/app.rs
Comment (jorgehermo9) [reply src/app.rs:550]: yes, that would fit

File: src/config/watcher.rs
Comment (amribm) [reply src/config/watcher.rs]: I just added the watch method for `Watcher`. that eliminates the dynamic dispatch. 
can u check if this is okay

File: src/config/watcher.rs
Comment (brittonhayes) [reply src/config/watcher.rs:158]: I'd vote against it being user controllable and hope for it to just have a sane default. I can't think of a scenario where I'd wanna toggle it!

Comment (jorgehermo9) [reply src/config/watcher.rs:158]: We found issues in openshift environments where the inotify watcher was constantly triggered and the file contents did not change at all. In this case, we would like to hash de contents to check if it really changed before reloading vector config. I think having sane defaults is good (for example, this would default to false), but there are situations where this is useful

File: src/config/watcher.rs
Comment (jorgehermo9) [inline src/config/watcher.rs:49]: I think you can get rid of this (In the other match arm too)

This behaviour was improved a few years ago  

Read about it here: https://github.com/rust-lang/rfcs/blob/master/text/2005-match-ergonomics.md


```suggestion
            Watcher::RecommendedWatcher(watcher) => {
```

File: src/config/watcher.rs
Comment (jorgehermo9) [inline src/config/watcher.rs:163]: I think you can take 
```
    watcher.add_paths(config_paths)?;
            Ok((watcher, receiver))
```
out of the match branches

```suggestion
    let watcher = match watcher_conf {
        WatcherConfig::RecommendedWatcher => {
            let recommended_watcher = recommended_watcher(sender)?;
            Watcher::RecommendedWatcher(recommended_watcher)
        }
        WatcherConfig::PollWatcher(interval) => {
            let config =
                notify::Config::default().with_poll_interval(Duration::from_secs(*interval));
            let poll_watcher = notify::PollWatcher::new(sender, config)?;
            Watcher::PollWatcher(poll_watcher)
        };

     watcher.add_paths(config_paths)?;
            Ok((watcher, receiver))

```

File: src/cli.rs
Comment (jorgehermo9) [inline src/cli.rs:170]: I think we should update this documentation to explain the polling watcher and when to use it

https://vector.dev/docs/administration/management/#automatic-reloading-on-configuration-change

But first lets see what maintainers think about documenting this

File: src/config/watcher.rs
Comment (brittonhayes) [reply src/config/watcher.rs:158]: Would there be a downside to always hashing the changes?

Comment (jorgehermo9) [reply src/config/watcher.rs:158]: Yes, as the docs I pointed states:

> By enabling this feature, performance will be significantly impacted as all files will need to be read and hashed at each poll_interval.

File: src/config/watcher.rs
Comment (amribm) [reply src/config/watcher.rs:34]: Since the `Watcher` is necessary for use `watch` method on `PollWatcher` and `RecommendWatcher` . removing that makes `watch` method is not defined error

File: src/config/watcher.rs
Comment (amribm) [reply src/config/watcher.rs:162]: yeah, this seems elegant. i'll do that

File: src/config/watcher.rs
Comment (jorgehermo9) [reply src/config/watcher.rs:34]: I think you can `use notify::RecommendedWatcher` and `use notify::PollingWatcher`, that have nothing related to what you're saying about the `Watcher` trait. The previous code imported those two structs

and indeed, maybe you need to `use notify::Watcher` to use the `watch` method

File: src/config/watcher.rs
Comment (jorgehermo9) [inline src/config/watcher.rs:50]: I think this would work
```suggestion
            Watcher::RecommendedWatcher(watcher) => {
```

File: src/config/watcher.rs
Comment (jorgehermo9) [inline src/config/watcher.rs:53]:
```suggestion
            Watcher::PollWatcher(watcher) => {
```

**Comment by pront** [comment]
Hey @amribm, thank you for this PR! 

* This requires a changelog entry. See https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md.
* I see there are a couple of open comment threads, is this ready for a review?

**Comment by amribm** [comment]
> * I see there are a couple of open comment threads, is this ready for a review?

Yes This is ready for review , @pront i need some confirmation from vector team for those comments. that's why i didn't resolved them.

File: src/cli.rs
Comment (amribm) [reply src/cli.rs:171]: where we need to add this in vector codebase?

File: src/app.rs
Comment (amribm) [reply src/app.rs:550]: Can i change this to `Inotify` as `Recommended`? since the `Inotify` is Platform dependent? @pront

Comment (pront) [reply src/app.rs:550]: Yes, we should try to keep this OS agnostic.

File: src/cli.rs
Comment (pront) [reply src/cli.rs:171]: A mention here along with the Rustdocs (that you already have) should suffice: https://vector.dev/docs/administration/management/#automatic-reloading-on-configuration-change

File: src/config/watcher.rs
Comment (pront) [inline src/config/watcher.rs:61]: I suppose this can be deleted

**Review by michaelcretzman — APPROVED** [review]

**Review by brittonhayes — APPROVED** [review]

File: src/cli.rs
Comment (jorgehermo9) [reply src/cli.rs:173]: I see that you resolved this comment @amribm but this documentation was not updated. I would prefer to update that before merging this

File: src/cli.rs
Comment (pront) [inline src/cli.rs:162]: The env var name is inconsistent with the `cli.cue` here. 
```suggestion
    #[arg(long, default_value = "inotify", env = "VECTOR_WATCH_CONFIG_METHOD")]
```

File: src/config/watcher.rs
Comment (pront) [reply src/config/watcher.rs:147]: Do you plan to add this option? Feel free to resolve this if not.

File: src/cli.rs
Comment (pront) [inline src/cli.rs:383]:
```suggestion
#[derive(clap::ValueEnum, Debug, Clone, Copy, PartialEq, Eq)]
pub enum WatchConfigMethod {
    /// Recommended watcher for the current OS, usually `inotify` for Linux-based systems.
    Recommended,
    /// Poll-based watcher, typically used for watching files on EFS/NFS-like network storage systems.
    /// The interval is determined by  [`RootOpts::watch_config_poll_interval_seconds`].
    Poll,
}
```

File: src/config/watcher.rs
Comment (pront) [inline src/config/watcher.rs:30]:
```suggestion
/// Refer to [`crate::cli::WatchConfigMethod`] for details.
pub enum WatcherConfig {
    /// Recommended watcher for the current OS.
    RecommendedWatcher,
    /// A poll-based watcher that checks for file changes at regular intervals.
    PollWatcher(u64),
}
```

**Review by pront — APPROVED** [review]
Left a few doc improvement suggestions, but otherwise this looks great.

File: changelog.d/poll-watcher-support.enhancement.md
Comment (pront) [inline changelog.d/poll-watcher-support.enhancement.md:1]: This can be a bit more expressive.

**Comment by pront** [comment]
> It seems that these comments are resolved [#21290 (comment)](https://github.com/vectordotdev/vector/pull/21290#discussion_r1759726163) but I dont see any documentation about this new option.
> 
> Shouldn't that be included in this PR? @pront

Let's include it in this PR 👍 cc @amribm

File: src/config/watcher.rs
Comment (amribm) [reply src/config/watcher.rs:148]: yeah. i don't have an idea how to make this to opt in. i feel like i'm not gonna add this.

File: src/cli.rs
Comment (amribm) [reply src/cli.rs:177]: Added documentation

File: website/content/en/docs/administration/management.md
Comment (jorgehermo9) [inline website/content/en/docs/administration/management.md:226]:
```suggestion
You can set the poll interval by setting flag `--watch-config-poll-interval-seconds`. which defaults to `30`.
```

File: website/content/en/docs/administration/management.md
Comment (jorgehermo9) [inline website/content/en/docs/administration/management.md:224]:
```suggestion
`poll` can used where the event listener won't work, eg. attaching config files by NFS/EFS, which will poll for file changes on certain interval.
```

Rephrasing this a little bit? maybe we can get a little help about technical writing in this file? @michaelcretzman

File: src/config/watcher.rs
Comment (pront) [inline src/config/watcher.rs:26]: I wonder if we can refactor `create_watcher` to use the `WatchConfigMethod` enum directly. AFAIK, due to clap parsing, the `WatchConfigMethod` needs to have plain enum variants.

**Review by pront — COMMENTED** [review]
Left one comment (TIOLI). Also, waiting for a documentation review. Otherwise, this is good to go 🚀

File: src/config/watcher.rs
Comment (amribm) [reply src/config/watcher.rs:26]: Since i want that enum to have values about the interval and i don't want to pass the interval in the function parameters. 

One thing we can do is we can eliminate is watch config method enum. we can make it as string and add the possible values for `clap` arguments. using that we can create the `WatcherConfig` enum

File: src/config/watcher.rs
Comment (pront) [inline src/config/watcher.rs:49]: nit: any reason to use this here vs at the top like the other imports?

**Review by pront — APPROVED** [review]

**Comment by pront** [comment]
I think we addressed all open questions here? cc @amribm and @jorgehermo9
If so, feel free to enqueue this for merging. Let me know if you need any input from me.

**Comment by jorgehermo9** [comment]
LGTM!

File: src/config/watcher.rs
Comment (amribm) [reply src/config/watcher.rs:49]: Actually we already have name watcher as enum. it was colliding with watcher from `notify`.
Renaming the import at the top level user might not able to understand what is `NotifyWathcher` and we need the methods for `notify::Watcher` on that specific scope. that's why i used Watcher on the specific scope

**Comment by pront** [comment]
Apologies, I didn't realize this got pushed out of the merge queue.

**Comment by amribm** [comment]
When this PR will merge @pront?

**Comment by jszwedko** [comment]
> When this PR will merge @pront?

It looks like there were cue formatting issues blocking merge as the failed CI check shows. I formatted and pushed though so I think this should merge soon if there are no other failing checks.

**Comment by pront** [comment]
Thanks again, this is a great contribution 🚀

**Comment by amribm** [comment]
Thank you guys, this is my first open-source contribution. thanks for making this possible. @jorgehermo9 @pront @brittonhayes

**Pull request merged**
